### ISSUE TITLE
Support SEED RL (SCALABLE AND EFFICIENT DEEP-RL )

### ISSUE THREAD

**Issue by findmyway** [JuliaReinforcementLearning/ReinforcementLearning.jl#62]
After reading the paper [SEED RL: SCALABLE AND EFFICIENT DEEP-RL WITH
ACCELERATED CENTRAL INFERENCE](https://arxiv.org/pdf/1910.06591.pdf), I think it is doable in Julia based on my current understanding. It may be our first attempt to support distributed RL.

![image](https://user-images.githubusercontent.com/5612003/77721364-9985b900-7025-11ea-8a54-b05699f8ce9a.png)

For actors, we already have [MultiThreadEnv](https://github.com/JuliaReinforcementLearning/ReinforcementLearningBase.jl/blob/3c460988dbacfeb57cdcbb042593ea27d62c2b3c/src/implementations/environments.jl#L43-L52) to run multiple environments in parallel. In the original paper, gRPC (streaming) is used to transfer observations/actions between actors and learners. I'm not sure it is implemented in [gRPC.jl](https://github.com/tanmaykm/gRPC.jl) or not. I guess using remote_call may not be very efficient. [ArrayChannels.jl](https://github.com/rohanmclure/ArrayChannels.jl) may also be helpful here.

For the learner part:

![image](https://user-images.githubusercontent.com/5612003/77721861-02216580-7027-11ea-9919-f70265f36bdf.png)

It is said that there are three types of threads:

1. Inference threads
2. Data prefetching threads
3. Training threads

The data prefetching threads are the easiest to understand. Simply fetch completed trajectories and update the prioritized buffer. Then generate training data (sending them to devices). It's just working as the `extract_experience` function in our package.

> The training thread (the main Python thread) takes the prefetched trajectories, computes
gradients using the training TPU cores and applies the gradients on the models of all TPU cores
(inference and training) synchronously. The ratio of inference and training cores can be adjusted for
maximum throughput and utilization. 

As it is said in the paper, inference and training steps are synchronous. So now we have to control the following parts:

1. the speed of generating training data (it will consume too much GPU memory), maybe a channel with limited length?
1. The ratio of inference and training. Just inference several batches after each training? I had tried to implement Ape-X before and found it difficult to control the speed.

Maybe we can start from the R2D2 on a single machine first? And then try SEED RL? 

(also cc @jbrea 😃  )

**Comment by drozzy** [comment]
ACME by DeepMind has the best distributed architecture in my opinion (after trying rllib, stable-baselines, and some others). Perhaps you can draw some inspiration from them?
https://github.com/deepmind/acme

**Comment by drozzy** [comment]
Imo this is inferior to ACME RL.

Also, google just published Menger https://ai.googleblog.com/2020/10/massively-large-scale-distributed.html

which uses some aspects of ACME RL. I think one downside of Menger is that it is tensorflow specific (afaik).

**Comment by drozzy** [comment]
Seed RL is by Brain Team and ACME is by DeepMind. Given that deepmind did pretty well with MuZero and Agent57 - I think they used acme for it.
